File: app/src/main/java/bogomolov/aa/fitrack/view/fragments/TracksListFragment.java
package bogomolov.aa.fitrack.view.fragments;


import android.content.Context;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.NavigationUI;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.transition.TransitionInflater;

import android.util.Log;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.view.animation.AnimationUtils;
import android.view.animation.LayoutAnimationController;
import android.widget.AdapterView;
import android.widget.Spinner;

import javax.inject.Inject;

import bogomolov.aa.fitrack.R;
import bogomolov.aa.fitrack.dagger.ViewModelFactory;
import bogomolov.aa.fitrack.databinding.FragmentTracksListBinding;
import bogomolov.aa.fitrack.core.DateUtils;
import bogomolov.aa.fitrack.core.model.Tag;
import bogomolov.aa.fitrack.view.TagResultListener;
import bogomolov.aa.fitrack.view.TracksPagedAdapter;
import bogomolov.aa.fitrack.viewmodels.TracksListViewModel;
import dagger.android.support.AndroidSupportInjection;


public class TracksListFragment extends Fragment implements TagResultListener {
    private TracksPagedAdapter adapter;
    private ActionMode actionMode;
    private Toolbar toolbar;

    @Inject
    ViewModelFactory viewModelFactory;

    private TracksListViewModel viewModel;
    private boolean spinnersCanClicked;
    private boolean canAnimate = true;


    private static final int FILTER_TODAY = 0;
    private static final int FILTER_WEEK = 1;
    private static final int FILTER_MONTH = 2;
    private static final int FILTER_SELECT = 3;

    @Override
    public void onAttach(Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(TracksListViewModel.class);
        FragmentTracksListBinding binding = DataBindingUtil.inflate(inflater, R.layout.fragment_tracks_list, container, false);
        binding.setViewModel(viewModel);
        binding.setLifecycleOwner(this);
        View view = binding.getRoot();


        toolbar = binding.toolbarTracksList;
        ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);

        NavController navController = Navigation.findNavController(getActivity(), R.id.nav_host_fragment);
        NavigationUI.setupWithNavController(toolbar, navController);

        RecyclerView recyclerView = binding.trackRecycler;
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        if (viewModel.tracksLiveData.getValue() == null)
            viewModel.updateTracks(DateUtils.getTodayRange());

        adapter = new TracksPagedAdapter(this);
        recyclerView.setAdapter(adapter);

        viewModel.tracksLiveData.observe(getViewLifecycleOwner(), tracks -> {
            Log.i("test", "loaded " + tracks.size());
            adapter.submitList(tracks);
                LayoutAnimationController animation = AnimationUtils.loadLayoutAnimation(TracksListFragment.this.getContext(), R.anim.track_item_layout_anim);
                recyclerView.setLayoutAnimation(animation);
        });

        recyclerView.getViewTreeObserver().addOnPreDrawListener(
                new ViewTreeObserver.OnPreDrawListener() {
                    @Override
                    public boolean onPreDraw() {
                        recyclerView.getViewTreeObserver().removeOnPreDrawListener(this);
                        startPostponedEnterTransition();
                        return true;
                    }
                });

        postponeEnterTransition();


        Spinner filterSpinner = binding.tracksTimeSpinner;
        filterSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                if (!spinnersCanClicked) return;
                switch (i) {
                    case FILTER_TODAY:
                        viewModel.updateTracks(DateUtils.getTodayRange());
                        break;
                    case FILTER_WEEK:
                        viewModel.updateTracks(DateUtils.getWeekRange());
                        break;
                    case FILTER_MONTH:
                        viewModel.updateTracks(DateUtils.getMonthRange());
                        break;
                    case FILTER_SELECT:
                        DateUtils.selectDatesRange(getChildFragmentManager(), dates -> viewModel.updateTracks(dates));
                        break;
                    default:
                        viewModel.updateTracks(DateUtils.getTodayRange());
                }
                Log.i("test","onItemSelected");
                canAnimate = true;
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });

        getActivity().getWindow().getDecorView().postDelayed(() -> spinnersCanClicked = true, 500);


        return view;
    }

    public void onLongClick() {
        if (actionMode == null)
            actionMode = toolbar.startActionMode(callback);
        else
            actionMode.finish();
    }

    private ActionMode.Callback callback = new ActionMode.Callback() {

        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            mode.getMenuInflater().inflate(R.menu.track_group_actions, menu);
            return true;
        }

        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            switch (item.getItemId()) {
                case R.id.menu_track_delete:
                    Log.i("test", "delete ids " + adapter.getSelectedIds());
                    viewModel.deleteTracks(adapter.getSelectedIds());
                    actionMode.finish();
                    break;
                case R.id.menu_track_tag:
                    TagSelectionDialog dialog = new TagSelectionDialog();
                    dialog.setTagResultListener(TracksListFragment.this);
                    dialog.show(getChildFragmentManager(), "TagSelectionDialog");
                    break;
            }
            return true;
        }

        public void onDestroyActionMode(ActionMode mode) {
            if (adapter != null) adapter.disableCheckMode();
            actionMode = null;
        }

    };

    @Override
    public void onTagSelectionResult(Tag tag) {
        if (tag != null) viewModel.setTag(tag, adapter.getSelectedIds());
        if (actionMode != null) actionMode.finish();
    }

}

File: app/src/main/java/bogomolov/aa/fitrack/view/fragments/TracksListFragment.kt
package bogomolov.aa.fitrack.view.fragments


import android.content.Context
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.NavController
import androidx.navigation.Navigation
import androidx.navigation.ui.NavigationUI
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

import android.view.ActionMode
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import android.view.animation.AnimationUtils
import android.view.animation.LayoutAnimationController
import android.widget.AdapterView
import android.widget.Spinner
import androidx.lifecycle.observe

import javax.inject.Inject

import bogomolov.aa.fitrack.R
import bogomolov.aa.fitrack.dagger.ViewModelFactory
import bogomolov.aa.fitrack.databinding.FragmentTracksListBinding
import bogomolov.aa.fitrack.core.*
import bogomolov.aa.fitrack.core.model.Tag
import bogomolov.aa.fitrack.view.TagResultListener
import bogomolov.aa.fitrack.view.TracksPagedAdapter
import bogomolov.aa.fitrack.viewmodels.TracksListViewModel
import dagger.android.support.AndroidSupportInjection


class TracksListFragment : Fragment(), TagResultListener {
    private lateinit var adapter: TracksPagedAdapter
    private var actionMode: ActionMode? = null
    private lateinit var toolbar: Toolbar

    @Inject
    internal lateinit var viewModelFactory: ViewModelFactory

    private lateinit var viewModel: TracksListViewModel
    private var spinnersCanClicked = false

    private val callback = object : ActionMode.Callback {

        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            mode.menuInflater.inflate(R.menu.track_group_actions, menu)
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
            return false
        }

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
            when (item.itemId) {
                R.id.menu_track_delete -> {
                    viewModel!!.deleteTracks(adapter.selectedIds)
                    actionMode!!.finish()
                }
                R.id.menu_track_tag -> {
                    val dialog = TagSelectionDialog()
                    dialog.tagResultListener = this@TracksListFragment
                    dialog.show(childFragmentManager, "TagSelectionDialog")
                }
            }
            return true
        }

        override fun onDestroyActionMode(mode: ActionMode) {
            if (adapter != null) adapter!!.disableCheckMode()
            actionMode = null
        }

    }

    override fun onAttach(context: Context) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        viewModel = ViewModelProvider(this, viewModelFactory).get(TracksListViewModel::class.java)
        val binding = DataBindingUtil.inflate<FragmentTracksListBinding>(inflater, R.layout.fragment_tracks_list, container, false)
        binding.viewModel = viewModel
        binding.lifecycleOwner = this
        val view = binding.root

        toolbar = binding.toolbarTracksList
        (activity as AppCompatActivity).setSupportActionBar(toolbar)

        val navController = Navigation.findNavController(activity!!, R.id.nav_host_fragment)
        NavigationUI.setupWithNavController(toolbar, navController)

        val recyclerView = binding.trackRecycler
        recyclerView.layoutManager = LinearLayoutManager(context)
        if (viewModel.tracksLiveData.value == null)
            viewModel.updateTracks(getTodayRange())

        adapter = TracksPagedAdapter(this)
        recyclerView.adapter = adapter

        viewModel.tracksLiveData.observe(viewLifecycleOwner) { tracks ->
            adapter.submitList(tracks)
            val animation = AnimationUtils.loadLayoutAnimation(this@TracksListFragment.context, R.anim.track_item_layout_anim)
            recyclerView.layoutAnimation = animation
        }

        recyclerView.viewTreeObserver.addOnPreDrawListener(
                object : ViewTreeObserver.OnPreDrawListener {
                    override fun onPreDraw(): Boolean {
                        recyclerView.viewTreeObserver.removeOnPreDrawListener(this)
                        startPostponedEnterTransition()
                        return true
                    }
                })

        postponeEnterTransition()

        val filterSpinner = binding.tracksTimeSpinner
        filterSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
                if (!spinnersCanClicked) return
                when (i) {
                    FILTER_TODAY -> viewModel.updateTracks(getTodayRange())
                    FILTER_WEEK -> viewModel.updateTracks(getWeekRange())
                    FILTER_MONTH -> viewModel.updateTracks(getMonthRange())
                    FILTER_SELECT -> selectDatesRange(childFragmentManager) { dates -> viewModel.updateTracks(dates) }
                    else -> viewModel.updateTracks(getTodayRange())
                }
            }

            override fun onNothingSelected(adapterView: AdapterView<*>) {}
        }

        activity!!.window.decorView.postDelayed({ spinnersCanClicked = true }, 500)

        return view
    }

    fun onLongClick() {
        if (actionMode == null)
            actionMode = toolbar.startActionMode(callback)
        else
            actionMode!!.finish()
    }

    override fun onTagSelectionResult(tag: Tag?) {
        if (tag != null) viewModel.setTag(tag, adapter.selectedIds)
        if (actionMode != null) actionMode!!.finish()
    }

    companion object {
        private val FILTER_TODAY = 0
        private val FILTER_WEEK = 1
        private val FILTER_MONTH = 2
        private val FILTER_SELECT = 3
    }

}
